Allow / in yang strings

tsyesika opened the pull request:
This should for the most part fix #716

Commit:
Fix #716 - allow / in yang strings

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -189,7 +204,19 @@ end
+         ret = ret .. self:take_while("[^%s;{}\"'/]")
+         if self:is_eof() then break end
+         if self:peek_n(2) == "/*" then break end
+         if self:peek() == "/" then
```
if self:check('/') and remove the consume below?

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -84,6 +92,13 @@ function Parser:check(expected)
    return false
 end
 
+function Parser:check_patten(patten)
```
check_pattern

wingo reviewed (commented):
Looks good, thank you!  Only a couple nits.

Commit:
#716 - Fix small nits brought up in review

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -84,6 +92,13 @@ function Parser:check(expected)
    return false
 end
 
+function Parser:check_pattern(patten)
```
"pattern" as the local variable please

tsyesika replied:
Ugh yeah, sorry. My ability to spell has always been somewhat lacking, I shall try to remember this one.

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -84,6 +92,13 @@ function Parser:check(expected)
 
+function Parser:check_pattern(patten)
+   if not self:is_eof() then
+      return self:peek():match(patten)
```
and of course updated here

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -189,7 +204,18 @@ end
+            ret = ret .. "/"
+         end
+      until not self:check_pattern("[^%s;{}\"'/]")
+      return ret
```
Since parse_string is on some hot paths, we should probably fix the quadratic algorithm before releasing; i.e. local ret = [], then table.insert the pieces as we parse them, then return `return table.concat(ret)`

tsyesika replied:
Ah yes, I did wonder about a lua equivalent of python's `"".join(list)`. Thanks for catching it.

wingo reviewed (commented):
Just a couple quick comments, sorry for not catching them the first time around

Commit:
#716 - Fix a few more nits

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -189,7 +204,18 @@ end
+         table.insert(ret, self:take_while("[^%s;{}\"'/]"))
+         if self:is_eof() then break end
+         if self:peek_n(2) == "/*" then break end
+         if self:check("/") then
```
Can you add a test for the string `foo//bar` which I assume should parse as just `foo`.  If that works, lgtm :)

tsyesika replied:
I sure can :)

Commit:
#716 - Add // comment handling in strings

Add // comment handling in the middle of unquoted strings and a
test to check for this situation.

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -189,7 +204,19 @@ end
+      local ret = {}
+      repeat
+         table.insert(ret, self:take_while("[^%s;{}\"'/]"))
+         if self:is_eof() then break end
```
The code looks good!  I am fine with it as-is.  Would you mind testing on a softwire-config with a big (1M) binding table and making sure parsing performance doesn't regress?  Thanks :)  That can be done after this goes in fwiw.  There are some speedups we could have here if that perf regresses.
